File: backend/meta/etl_task.py
"""
ETL任务实体
ETLTask连接Datasource和Model，不依赖Domain
"""
from typing import Optional, Any
from datetime import date
import json


class ETLTask:
    """ETL任务实体 - 连接Datasource和Model"""
    
    def __init__(
        self,
        id: int,
        name: str,
        sourceDatasourceId: int,
        targetModelId: int,
        description: Optional[str] = None,
        status: str = "inactive",
        schedule: Optional[str] = None,
        config: Optional[Any] = None,
        createdAt: Optional[str] = None,
        updatedAt: Optional[str] = None,
        lastRun: Optional[str] = None,
        nextRun: Optional[str] = None
    ):
        self.id = id
        self.name = name
        self.sourceDatasourceId = sourceDatasourceId
        self.targetModelId = targetModelId
        self.description = description
        self.status = status  # inactive, active, running, paused, error
        self.schedule = schedule
        self.config = config
        self.createdAt = createdAt or date.today().isoformat()
        self.updatedAt = updatedAt or date.today().isoformat()
        self.lastRun = lastRun
        self.nextRun = nextRun
    
    def to_dict(self) -> dict:
        """转换为字典"""
        config = self.config
        if isinstance(config, str):
            try:
                config = json.loads(config)
            except json.JSONDecodeError:
                pass
        
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "sourceDatasourceId": self.sourceDatasourceId,
            "targetModelId": self.targetModelId,
            "status": self.status,
            "schedule": self.schedule,
            "config": config,
            "createdAt": self.createdAt,
            "updatedAt": self.updatedAt,
            "lastRun": self.lastRun,
            "nextRun": self.nextRun
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'ETLTask':
        """从字典创建实例"""
        return cls(
            id=data["id"],
            name=data["name"],
            sourceDatasourceId=data["sourceDatasourceId"],
            targetModelId=data["targetModelId"],
            description=data.get("description"),
            status=data.get("status", "inactive"),
            schedule=data.get("schedule"),
            config=data.get("config"),
            createdAt=data.get("createdAt"),
            updatedAt=data.get("updatedAt"),
            lastRun=data.get("lastRun"),
            nextRun=data.get("nextRun")
        )
    
    def activate(self) -> 'ETLTask':
        """激活任务"""
        self.status = "active"
        self.updatedAt = date.today().isoformat()
        return self
    
    def pause(self) -> 'ETLTask':
        """暂停任务"""
        self.status = "paused"
        self.updatedAt = date.today().isoformat()
        return self
    
    def start(self) -> 'ETLTask':
        """启动任务"""
        self.status = "running"
        self.updatedAt = date.today().isoformat()
        return self
    
    def complete(self, last_run: Optional[str] = None) -> 'ETLTask':
        """完成任务"""
        self.status = "active"
        self.lastRun = last_run or date.today().isoformat()
        self.updatedAt = date.today().isoformat()
        return self
    
    def error(self) -> 'ETLTask':
        """标记任务错误"""
        self.status = "error"
        self.updatedAt = date.today().isoformat()
        return self
    
    def __repr__(self):
        return f"ETLTask(id={self.id}, name='{self.name}', status='{self.status}')"

File: backend_java/src/main/java/com/jianmo/platform/entity/ETLTask.java
package com.jianmo.platform.entity;

import com.jianmo.platform.common.BaseEntity;
import javax.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.ToString;

import java.util.ArrayList;
import java.util.List;


@Entity
@Table(name = "etl_tasks")
@EqualsAndHashCode(callSuper = true)
@Data
@NoArgsConstructor
@AllArgsConstructor
public class ETLTask extends BaseEntity {

    @Column(nullable = false)
    private String name;

    @Column(name = "source_datasource_id")
    private Long sourceDatasourceId;

    @Column(name = "target_model_id")
    private Long targetModelId;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "task_status")
    private String status = "inactive";

    @Column(columnDefinition = "TEXT")
    private String schedule;

    @Column(columnDefinition = "TEXT")
    private String config;

    @Column(name = "last_run")
    private String lastRun;

    @Column(name = "next_run")
    private String nextRun;

    @OneToMany(mappedBy = "task", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @ToString.Exclude
    private List<ETLLog> logs = new ArrayList<>();

    public ETLTask(String name, Long sourceDatasourceId, Long targetModelId) {
        this.name = name;
        this.sourceDatasourceId = sourceDatasourceId;
        this.targetModelId = targetModelId;
    }

    public void addLog(ETLLog log) {
        logs.add(log);
        log.setTaskId(this.getId());
    }

    public boolean activate() {
        if ("inactive".equals(status) || "paused".equals(status)) {
            this.status = "active";
            return true;
        }
        return false;
    }

    public boolean pause() {
        if ("active".equals(status)) {
            this.status = "paused";
            return true;
        }
        return false;
    }

    public boolean start() {
        if ("active".equals(status)) {
            this.status = "running";
            return true;
        }
        return false;
    }

    public boolean complete() {
        if ("running".equals(status)) {
            this.status = "active";
            this.lastRun = java.time.LocalDate.now().toString();
            return true;
        }
        return false;
    }

    public boolean markError() {
        if ("running".equals(status)) {
            this.status = "error";
            return true;
        }
        return false;
    }

    public boolean isValid() {
        return name != null && !name.isEmpty() &&
               sourceDatasourceId != null &&
               targetModelId != null;
    }
}
